### PR TITLE
Skip part of CompositeFormatterSuccess in OS X

### PR DESCRIPTION
**Pull request by rgs1**
Apparently %E is special in OS X?

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

**Comment by rgs1** [comment]
cc: @zasweq

**Comment by rgs1** [comment]
I get:

```
[ RUN      ] SubstitutionFormatterTest.CompositeFormatterSuccess
TestRandomGenerator running with seed 173593824
test/common/formatter/substitution_formatter_test.cc:2223: Failure
Expected equality of these values:
  "%E4n"
  formatter.format(request_header, response_header, response_trailer, stream_info, body)
    Which is: "4n"
Stack trace:
  0x106eed9f1: Envoy::Formatter::(anonymous namespace)::SubstitutionFormatterTest_CompositeFormatterSuccess_Test::TestBody()
  0x10bbf83f4: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x10bbcda9b: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x10bbcd9d3: testing::Test::Run()
  0x10bbce9f7: testing::TestInfo::Run()
... Google Test internal frames ...
```

**Comment by rgs1** [comment]
Ah so strftime() behaves differently in OS X then:

Godbolt: https://godbolt.org/z/WeeoKf

OS X:

```
$ cat strftime.cc
#include <ctime>
#include <iostream>
#include <locale>

int main()
{
  std::time_t t = std::time(nullptr);
  char mbstr[100];
  if (std::strftime(mbstr, sizeof(mbstr), "%E4n", std::localtime(&t))) {
    std::cout << mbstr << '\n';
  }
}
$ g++ -Wall -std=c++17 strftime.cc -o strftime
$ ./strftime
4n
```

**Comment by mattklein123** [comment]
Please merge main to fix OSX CI.

/wait

**Comment by rgs1** [comment]
> Ah so strftime() behaves differently in OS X then:
> 
> Godbolt: https://godbolt.org/z/WeeoKf
> 
> OS X:
> 
> ```
> $ cat strftime.cc
> #include <ctime>
> #include <iostream>
> #include <locale>
> 
> int main()
> {
>   std::time_t t = std::time(nullptr);
>   char mbstr[100];
>   if (std::strftime(mbstr, sizeof(mbstr), "%E4n", std::localtime(&t))) {
>     std::cout << mbstr << '\n';
>   }
> }
> $ g++ -Wall -std=c++17 strftime.cc -o strftime
> $ ./strftime
> 4n
> ```

It's actually not `strftime()` what's used, it's `absl::FormatTime()` called from:

https://github.com/envoyproxy/envoy/blob/master/source/common/common/utility.cc#L204

Though `absl::FormatTime()` follows `strftime()` formatting options but adds extensions:

https://github.com/abseil/abseil-cpp/blob/master/absl/time/time.h#L1219

Not sure why this works in CI.

File: test/common/formatter/substitution_formatter_test.cc [hunk 1]
```diff
@@ -2213,7 +2213,7 @@ TEST(SubstitutionFormatterTest, CompositeFormatterSuccess) {
         "%%|%%123456000|1522796769%%123|1%%1522796769",
         formatter.format(request_header, response_header, response_trailer, stream_info, body));
   }
-#ifndef WIN32
+#if !defined(WIN32) && !defined(__APPLE__)
```
Comment (mattklein123) [inline test/common/formatter/substitution_formatter_test.cc:2216]: Can you add a comment on why the ifdef is here on both platforms? I think some of the research you did would be helpful to the next person that comes along? Thank you.

/wait

Comment (rgs1) [reply test/common/formatter/substitution_formatter_test.cc:2216]: Yep -- doing some more local testing with Abseil  and then I'll push something.

**Comment by rgs1** [comment]
Ok, yes Abseil behaves differently in OS X:

```
$ cat test_abseil.cc
#include <iostream>
#include <string>
#include <vector>

#include "absl/time/time.h"

int main() {
  absl::TimeZone lax;
  auto t = absl::FromTimeT(time(nullptr));
  std::cout << absl::FormatTime("%E4n", t, lax) << "\n";

  return 0;
}
$ g++ -Wall -std=c++17 -framework CoreFoundation  -labsl_time -labsl_time_zone test_abseil.cc -o test_abseil
[05:58:51] ~/src/cpp $ ./test_abseil.cc
-bash: ./test_abseil.cc: Permission denied
$ ./test_abseil
4n
```

I'll add a comment.

File: test/common/formatter/substitution_formatter_test.cc (hunk 1 above)
Comment (rgs1) [reply test/common/formatter/substitution_formatter_test.cc:2220]: Done.

**Comment by rgs1** [comment]
Failures are integration tests:

```
//test/integration:protocol_integration_test                             FAILED in 10 out of 10 in 75.7s
  Stats over 10 runs: max = 75.7s, min = 61.8s, avg = 69.0s, dev = 4.9s
```

and

```
[  FAILED  ] Protocols/ProtocolIntegrationTest.ContinueAfterLocalReply/IPv4_HttpDownstream_Http2Upstream, where GetParam() = 12-byte object <00-00 00-00 00-00 00-00 01-00 00-00>
[  FAILED  ] Protocols/ProtocolIntegrationTest.ContinueAfterLocalReply/IPv6_Http2Downstream_HttpUpstream, where GetParam() = 12-byte object <01-00 00-00 01-00 00-00 00-00 00-00>


```

**Comment by mattklein123** [comment]
Merge main to fix OSX CI (and also format needs fixing).

/wait

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**